added contracts entry point

ckatsube opened the pull request:
<!-- Provide a summary of your changes in the Pull Request Title above. -->
<!-- If this is a work in progress (not yet ready to be merged), make this a draft pull request. -->

## Motivation and Context

<!-- Why is this pull request required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here: -->
<!-- https://docs.github.com/en/github/managing-your-work-on-github/managing-your-work-with-issues-and-pull-requests/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->

See #778 for an introduction to the motivation and context.
778 was closed due to the approach where decompiling the file into its AST changed the line numbers, resulting in unhelpful error tracebacks.

## Your Changes

<!-- Describe your changes here. -->

**Description**:

Added `__main__.py` for the contracts module which takes in the path to Python script to be executed as main, additional module names to be checked, and whether the main module should be checked.

Used a pseudo module to replace the main entry point for contract checking the main module

Alternatives:
- could have set checking main to be permanently false and added the pseudo module to the list of extra imports and used the cli flag to determine whether or not to add the pseudo module to the list or not
- ~~can just have one layer of execs and manually pass in the context dictionaries into the before and after execs.~~

**Type of change** (select all that apply):

<!-- Put an `x` in all the boxes that apply. -->
<!-- Remove any lines that do not apply. -->

- [x] New feature (non-breaking change which adds functionality)

## Testing

<!-- Please describe in detail how you tested this pull request. -->
<!-- This can include tests you added and manual testing. -->

Confirmed test suite still runs.
Used two modules (not part of PR) to test the successful assertion error on contract violation for both a main and extra module.

## Questions and Comments (if applicable)

<!-- Ask any questions you have for the maintainers of this project regarding this PR. -->
<!-- Please describe the steps you have already taken to find the answer to your question. -->
<!-- This will ensure that we can give you clear and relevant advice. -->
<!-- If you have additional comments add them here as well. -->

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the CI tests have passed. <!-- (check after opening pull request) -->
- [x] I have reviewed the test coverage changes reported on Coveralls. <!-- (check after opening pull request) -->

ckatsube commented:
Changelog:
- Added CLI sections on the `contracts/index.md` documentation page.
- Added proper exit codes
- Refactored out `astroid` with vanilla Python processing
- Updated the name for the flag which disables contract checking for the main module
- Refactored imports
- Updated docstring to be more accurate

ckatsube commented on python_ta/contracts/__main__.py:
```diff
@@ -0,0 +1,85 @@
+
+
+@click.command()
+@click.argument("main", type=click.File(mode="r"))
```
Click arguments don't have a `help` parameter. It seems like `click` suggests including documentation for arguments inside the docstring, as the docstring also appears when using the `--help` flag.
I think I can convert the `EXTRA_MOD_NAMES` to be a flag instead (similar to Pytest's `--ignore=path` flag), but I can't see `FILE` not being an argument.

Source: https://click.palletsprojects.com/en/8.0.x/documentation/#documenting-arguments

david-yz-liu commented on docs/contracts/index.md:
````diff
@@ -26,6 +26,11 @@ if __name__ == '__main__':
     divide(10, 0)  # Preconditions on y violated
 ```
 
+```{note}
````
You can remove this note, I don't want to distract from the main flow of the page here. It's good to show an example, but this should be done in the CLI section below.

david-yz-liu commented on docs/contracts/index.md:
````diff
@@ -67,6 +72,20 @@ Finally, you can set the `DEBUG_CONTRACTS` constant to `True` to enable debuggin
 .. autodata:: python_ta.contracts.DEBUG_CONTRACTS
 ```
 
+## CLI
````
Don't use the acronym here, Instead, use "Command-line interface".

david-yz-liu commented on docs/contracts/index.md:
````diff
@@ -486,3 +505,10 @@ The above example outputs the following
 >>> Did not raise AssertionError
 ```
+
+### Contracts CLI Limitations
````
Don't make this a separate section. Instead, you can make this a "note" in the CLI section.

david-yz-liu commented on docs/contracts/index.md:
```diff
@@ -67,6 +72,20 @@ Finally, you can set the `DEBUG_CONTRACTS` constant to `True` to enable debuggin
 
+## CLI
+
+The `python_ta.contracts` CLI can execute a file as `__main__` with contracts enabled. This option
```
The second sentence ("This option...") is redundant with the "limitations" section below. Remove this sentence.

david-yz-liu commented on python_ta/contracts/__main__.py:
```diff
@@ -0,0 +1,109 @@
+
+    main_lineno = _find_main_lineno(lines)
+
+    def _formatted_traceback() -> str:
```
Don't make this a nested function, move it to the global scope.

david-yz-liu commented on python_ta/contracts/__main__.py:
```diff
@@ -0,0 +1,109 @@
+
+@click.command()
+@click.argument("file", type=click.File(mode="r"))
+@click.argument("extra-mod-names", nargs=-1)
```
Yeah, you can make this an option. You can make it a comma-separated list of module names.

david-yz-liu commented on python_ta/contracts/__main__.py:
```diff
@@ -0,0 +1,109 @@
+    return 0
+
+
+def _has_main_check(line: str):
```
Add a return type here

david-yz-liu commented on python_ta/contracts/__main__.py:
```diff
@@ -0,0 +1,109 @@
+def _has_main_check(line: str):
+    if line.strip() == "":
+        return False
+    keyword, *condition = line.split()
```
To be safe, also check that the line starts at the top level (i.e., there's no indentation).

david-yz-liu replied:
@ckatsube I'm not sure if you missed this comment, or had a question about it?

ckatsube replied:
The change I made was a few more lines down (line 103 has `line.startswith(keyword)`).

david-yz-liu commented on docs/contracts/index.md:
```diff
@@ -67,6 +72,20 @@ Finally, you can set the `DEBUG_CONTRACTS` constant to `True` to enable debuggin
+requires there to be an `if __name__ == "__main__":` so PythonTA knows when and where to
+start tracking contracts from.
+
+The syntax for this command is: `python -m python_ta.contracts [OPTIONS] FILE [EXTRA_MOD_NAMES]`
```
You don't need to specify the options here. Instead, show just the basic usage, and then say that the user can run `python_ta.contracts --help` for information about the command-line options.

david-yz-liu replied:
Oh and say that the user might need to use `python3` instead of `python` depending on their system.

david-yz-liu commented on docs/contracts/index.md:
```diff
@@ -67,6 +72,20 @@ Finally, you can set the `DEBUG_CONTRACTS` constant to `True` to enable debuggin
+where `FILE` is the main script, `EXTRA_MOD_NAMES` are the names of imported modules to also be checked,
+and `--no-decorate-main`, the only available flag, is whether to decorate the main script.
+
+The utmost basic usage of this command is with `python -m python_ta.contracts FILE`.
```
This sentence should be merged with the previous paragraph.

david-yz-liu reviewed (commented):
@ckatsube Great, I left some more comments. Also please update the Changelog to add an entry for this update.

ckatsube commented:
Changelog: 
- Changed extra module argument to option (had to make it a multiple option since options don't support `nargs=-1`)
- Refactored traceback formatting helper to be global instead of local
- Main if block check now includes making sure the statement is at the module level
- Consolidated some text on the contract's documentation page to be more together and specific.
